internal/cmd/grpc.go

```diff
@@ -262,6 +265,32 @@ func NewGRPCServer(
 		logger.Debug("cache enabled", zap.Stringer("backend", cacher))
 	}
 
+	sinks := make([]auditsink.AuditSink, 0)
+	var sinksString string
+
+	if cfg.Audit.Sinks.LogFile.Enabled {
+		logFileSink, err := logfile.NewLogFileSink(logger, cfg.Audit.Sinks.LogFile.FilePath)
+		if err != nil {
+			return nil, fmt.Errorf("opening file at path: %s", cfg.Audit.Sinks.LogFile.FilePath)
+		}
+
+		sinks = append(sinks, logFileSink)
+		sinksString += fmt.Sprintf("%s,", logFileSink.String())
+	}
+
+	// Based on audit sink configuration from the user, provision the audit sinks and add them to a slice,
+	// and if the slice has a non-zero length, add the audit sink interceptor.
+	if len(sinks) > 0 {
+		publisher := auditsink.NewPublisher(logger, cfg.Audit.Buffer.Capacity, sinks, 2*time.Minute)
+		interceptors = append(interceptors, middlewaregrpc.AuditSinkUnaryInterceptor(logger, publisher))
+		logger.Debug(fmt.Sprintf("sinks are enabled with buffer size %d", cfg.Audit.Buffer.Capacity), zap.String("sinks", strings.TrimSuffix(sinksString, ",")))
```

Comment:
Nit: `logger.Debugf` would prevent the need for `fmt.Sprintf`

Reply:
One little formula I use with structured loggers is: if I feel the need to format a string for the log message, how can I adjust it so that the message is constant and the dynamic parts are fields in the message instead.

For example here (also note that zap appears to have a neat generic `zap.Stringers` now):
```go
		logger.Debug("sinks enabled", zap.Stringers("sinks", sinks), zap.Int("buffer_capacity", cfg.Audit.Buffer.Capacity))
```
It may be that `zap.Stringers` needs the generic argument filled in, in which case it would be `zap.Stringers[auditsink.AuditSink](sinks)` here.

This has the added benefit that I can filter in a logging system for a constant well known message and extract the capacity using the well known field key `buffer_capacity` if needs be. Rather than having to parse a string.

Reply:
Also, they hide the `f` prefixed methods in the `zap.SuggeredLogger` because formatting strings requires interpolation as is more inefficient. They want you to use more fields than formatting with structured logging. Thats a nice thing about zap in general, as it is making it hard for you to do the expensive thing on purpose. Encouraging you to lean into fields more.

Reply:
@GeorgeMac I like the `Stringers` generic. Takes the manual constructing of the string out of my code!